partiql-parser/src/main/kotlin/org/partiql/parser/internal/V1PartiQLParserDefault.kt

```diff
@@ -0,0 +1,2195 @@
+            }
+        }
+    }
+
+    /**
+     * Catches Parser errors (malformed syntax) and throws a [PartiQLParserException]
+     */
+    private class ParseErrorListener(private val listener: PErrorListener) : BaseErrorListener() {
+
+        private val rules = GeneratedParser.ruleNames.asList()
+
+        @Throws(PartiQLParserException::class)
+        override fun syntaxError(
+            recognizer: Recognizer<*, *>?,
+            offendingSymbol: Any,
+            line: Int,
+            charPositionInLine: Int,
+            msg: String,
+            e: RecognitionException?,
+        ) {
+            if (offendingSymbol is Token) {
+                val rule = e?.ctx?.toString(rules) ?: "UNKNOWN" // TODO: Do we want to display the offending rule?
+                val token = offendingSymbol.text
+                val tokenType = GeneratedParser.VOCABULARY.getSymbolicName(offendingSymbol.type)
+                val location = org.partiql.spi.SourceLocation(line.toLong(), charPositionInLine + 1L, token.length.toLong())
+                val error = PErrors.unexpectedToken(location, tokenType, null)
+                listener.report(error)
+            } else {
+                throw IllegalArgumentException("Offending symbol is not a Token.")
+            }
+        }
+    }
+
+    /**
+     * A wrapped [GeneratedParser] to allow thread interruption during parse.
+     */
+    internal class InterruptibleParser(input: TokenStream) : GeneratedParser(input) {
+        override fun enterRule(localctx: ParserRuleContext?, state: Int, ruleIndex: Int) {
+            if (Thread.interrupted()) {
+                throw InterruptedException()
+            }
+            super.enterRule(localctx, state, ruleIndex)
+        }
+    }
+
+    /**
+     * This token stream creates [parameterIndexes], which is a map, where the keys represent the
+     * indexes of all [GeneratedLexer.QUESTION_MARK]'s and the values represent their relative index amongst all other
+     * [GeneratedLexer.QUESTION_MARK]'s.
+     */
+    internal open class CountingTokenStream(tokenSource: TokenSource) : CommonTokenStream(tokenSource) {
+        // TODO: Research use-case of parameters and implementation -- see https://github.com/partiql/partiql-docs/issues/23
+        val parameterIndexes = mutableMapOf<Int, Int>()
+        private var parametersFound = 0
+        override fun LT(k: Int): Token? {
+            val token = super.LT(k)
+            token?.let {
+                if (it.type == GeneratedLexer.QUESTION_MARK && parameterIndexes.containsKey(token.tokenIndex).not()) {
+                    parameterIndexes[token.tokenIndex] = ++parametersFound
+                }
+            }
+            return token
+        }
+    }
+
+    /**
+     * Translate an ANTLR ParseTree to a PartiQL
+     */
+    @OptIn(PartiQLValueExperimental::class)
+    private class Visitor(
+        private val tokens: CommonTokenStream,
+        private val locations: SourceLocations.Mutable,
```

Comment:
not for this PR, but I would like to put source locations class under the microscope..